README.md
[//]: # (Image References)

[image1]: ./examples/placeholder.png "Model Visualization"
[image2]: ./examples/placeholder.png "Grayscaling"
[image3]: ./examples/placeholder_small.png "Recovery Image"
[image4]: ./examples/placeholder_small.png "Recovery Image"
[image5]: ./examples/placeholder_small.png "Recovery Image"
[image6]: ./examples/placeholder_small.png "Normal Image"
[image7]: ./examples/placeholder_small.png "Flipped Image"

# Behavioral Cloning Project [P3]
### Summary by Michael Berner, Student @ Udacity Self Driving Car NanoDegree 2018

[![Udacity - Self-Driving Car NanoDegree](https://s3.amazonaws.com/udacity-sdc/github/shield-carnd.svg)](http://www.udacity.com/drive)

Stuttgart, May 1st 2018

## Task / Problem description
*In this project, you will use what you've learned about deep neural networks and convolutional neural networks to clone driving behavior. You will train, validate and test a model using Keras. The model will output a steering angle to an autonomous vehicle.*

*We have provided a simulator where you can steer a car around a track for data collection. You'll use image data and steering angles to train a neural network and then use this model to drive the car autonomously around the track.* 

![Udacity_Simulator](./examples/Udacity_Simulator.png "Udacity Simulator")

*We also want you to create a detailed writeup of the project. Check out the [writeup template](https://github.com/udacity/CarND-Behavioral-Cloning-P3/blob/master/writeup_template.md) for this project and use it as a starting point for creating your own writeup. The writeup can be either a markdown file or a pdf document.*


**The goals / steps of this project are the following:**

* Use the simulator to collect data of good driving behavior
* Build, a convolution neural network in Keras that predicts steering angles from images
* Train and validate the model with a training and validation set
* Test that the model successfully drives around track one without leaving the road
* Summarize the results with a written report

The full original project description is located at [Udacity's GitHub repository](https://github.com/udacity/CarND-Behavioral-Cloning-P3).

## Project Writeup & Rubric Points

In this section, I will consider the [rubric points](https://review.udacity.com/#!/rubrics/432/view) individually and describe how I addressed each point in my implementation. 


### Files Submitted & Code Quality

#### 1. Required files
*Submission includes all required files and can be used to run the simulator in autonomous mode*

My project includes the following files:

* **model.py** containing the script to create and train the model
* **drive.py** for driving the car in autonomous mode
* **model.h5** containing a trained convolution neural network 
* **README.md** summarizing the results (write_up.md was replaced with this readme file)

All files can be downloaded easily from my GitHub repository using the command

```
git clone https://github.com/thelukasssheee/CarND-Behavioral-Cloning-P3.git
```

#### 2. Submission includes functional code
Using the Udacity provided simulator and my ``drive.py`` file, the car can be driven autonomously around the track by executing ``python drive.py model.h5``

For additional fun and challenge, I tweaked the ``drive.py``to run at 30 mph instead of 8 mph. With the provided neural network, the car is still able to remain on the tarmac.

However, it is necessary to activate the Anaconda development environment carnd-term1 to have all necessary Python functions available with the correct version. ``source activate carnd-term1``

#### 3. Submission code is usable and readable

The ``model.py`` file contains the code for training and saving the convolution neural network. The file shows the pipeline I used for training and validating the model, and it contains comments to explain how the code works.

The code is structured in such a way, that it can read the training data which I created alongside with the Udacity data. This is indicated by the array ``use_datasets = [0,1,2]``. If you want to run the code with the originally provided data by Udacity in precisely the form it was provided, you need to uncomment the line ``# use_datasets = [3]`` and comment the line ``use_datasets = [0,1,2]``.

```
### Settings
use_datasets = [0,1,2]  # Array to select which datasets to process
# use_datasets = [3]      # Array to select which datasets to process
nb_epochs = 7           # Number of epochs for neural network training
batch_sz = 32           # Batch size for neural network
test_sz = 0.20          # Fraction of images to use for test set
steer_corr = 0.00       # Steering correction value (left, right camera)
dropout_keep_rate = 0.7 # Dropout rate to keep neurons

### List of available datasets
datasets = np.array([   'data/Udacity/driving_log.csv', \
                        'data/T1_Regular/driving_log.csv', \
                        'data/T1_OtherDir/driving_log.csv', \
                        'data/driving_log.csv'], \
                        dtype='str')
datasets = datasets[use_datasets]
```

### Model Architecture and Training Strategy

#### 1. An appropriate model architecture has been employed

My model consists of a convolution neural network, derived from a powerful network by [nVidia](https://devblogs.nvidia.com/) which was described in [their blog](https://devblogs.nvidia.com/parallelforall/deep-learning-self-driving-cars). 

The network starts with normalization and image cropping.

These steps are followed by several convolutions, each followed by a non-linearity through a ReLU activation function.  

#### 2. Attempts to reduce overfitting in the model

I experimented a bit with dropout layers, since I gathered a lot of positive experiences from the last project. 

However, for this neural network, the benefit of introducing dropouts was very limited. Almost no accuracy gain could be identified. The reason for this is most likely, that my neural network was trained on almost 50000 images, helping significantly to reduce the risk of overfitting.

The model was trained and validated on different data sets to ensure that the model was not overfitting (code line 25-28 combined with line 16).

The model was tested by running it through the simulator and ensuring that the vehicle could stay on the track.

#### 3. Model parameter tuning

The model used an adam optimizer, so the learning rate was not tuned manually (model.py line 138).

However, an additional parameter was assessed in detail. It was proposed, to use the left and right camera images as pseudo-center images for the training algorithm. The idea was to add some artificial steering angle to the left and right based camera views.

I tweaked this parameter in quite a wide range with a surprising result: reducing the additionally introduced steering angle back to 0.00 was yielding the best and most stable result.

| Additional steering angle  |     Validation loss | Result 8mph | Result 20 mph | Result 30 mph |
|:-------------------------:|:---------------------:|:-----------:|:------:|:----:| 
| 0.2          			| 0.0154 		| PASS | PASS | FAIL |
| 0.4   					| 0.0357  	| PASS | barely PASS | FAIL |
| 0.1			 			| 0.00101  	| PASS | PASS | FAIL |
| 0.05						| 0.0097   	| PASS | PASS | FAIL |
| 0.00						| 0.0092		| PASS | PASS | PASS |

#### 4. Appropriate training data

Training data was chosen to keep the vehicle driving on the road. 

* Udacity training data
* 1 lap track 1 counter-clockwise
* 1 lap track 1 clockwise

Most of this data consisted of center lane driving. 

Originally, I planned to add some image data consisting of recovery maneuvers from the left or right side of the road. This was not at all necessary! By introducing left and right camera images to the overall pool and augmenting the avilable image data (flip around vertical axis), the overall result was becoming very stable indeed. 

### Model Architecture and Training Strategy

#### 1. Solution Design Approach

The overall strategy for deriving a model architecture was to generate a ``model.py`` file, which was able to run locally on my iMac 2011. 

Therefore, I started with a very basic network (easy to compute) and made the script ready to handle large amounts of image data. First attemps to introduce pickled data sets (faster to load) were discarded, because I could already see memory issues coming up on the horizon... 

I increased the amount of images available from a few thousand up to 50000 in the end. However, the basic model was still not able to keep my car on the road. In the version 5, after several seconds, the car leaves the bridge and takes a refreshing bath in the lake... 

Below you can see a lot of failed attempts:

Model v1: Enjoying a bath... (loss 2519)

![Model v1](./examples/record1.gif "Model v1: Is it a boat?")

Model v2: Additional training lap CCW introduced, biased CCW (loss 1681)

![Model v2](./examples/record2.gif "Model v2: Donuts CCW")

Model v3: Additional training lap CCW & long CW, now biased CW (loss 1901)

![Model v3](./examples/record3.gif "Model v3: Donuts CW")

Model v5: Normalization, image cropping, left+right cam images, flipped images / augmentation (loss 0.1884)

![Model v5](./examples/record5.gif "Model v5: Heading in the right direction!")



Then I decided that it was time move on to a more complex model. Since my iMac couldn't handle this computation, I ran it on Amazon's Web Services "Elastic Computing 2" machine with a GPU attached. 

#### 2. Final Model Architecture

As already mentioned, the final model architecture was derived from the [nVidia blog](https://devblogs.nvidia.com/parallelforall/deep-learning-self-driving-cars).

The nVidia network, which was the basis for my CNN, is depicted in the following image:
![nVidia CNN](./examples/cnn-architecture-624x890.png "nVidia CNN")

My final model architecture (model.py lines 18-24) consists of a convolution neural network with the layers and layer sizes explained in the table below.

| Layer		     			|     Description | 
|:-------------------------:|:---------------------------------------------:| 
| Input         			| 160x320x3 RGB image files (jpg) | 
| Lambda Normalization   | RGB / 255.0 - 0.5       |
| Cropping 2D 			|	skip upper 70 and lower 25 pixels, output: 65x320x3   |
| Convolution 2D 5x5 with ReLU  | 24 filters, 2x2 stride, valid padding |
| Convolution 2D 5x5 with ReLU  | 36 filters, 2x2 stride, valid padding |
| Convolution 2D 5x5 with ReLU  | 48 filters, 2x2 stride, valid padding |
| Convolution 2D 3x3 with ReLU  | 64 filters, no stride, valid padding |
| Convolution 2D 3x3 with ReLU  | 64 filters, no stride, valid padding |
| Flatten					       | 	                                     |
| Fully connected (Dense)		| Output dimension 100x100               |
| Dropout 						| Keep probability 0.7                   |
| Fully connected (Dense)		| Output dimension 50x50                 |
| Fully connected (Dense) 		| Output dimension 10x10                 |
| Fully connected (Dense)		| Output dimension 1x1                   |

The final output layer is only of dimension 1x1, since it is a sequential neural network, which is providing only one information: the steering angle for the car simulator. 

In the ``model.py`` file, these functions can be found from line 120 through line 136. 


#### 3. Creation of the Training Set & Training Process

To capture good driving behavior, I first recorded one lap on track one using center lane driving. Afterwards, my model was heavily biased towards counter-clockwise driving. 

Model v2: Additional training lap CW introduced, biased CW (loss 1681)

![Model v2](./examples/record2.gif "Model v2: Donuts CW")

I then added images from the left and right camera angle, to artificially generate more data for the CNN where the vehicle is close the sides of the road alongside with an artificial steering command (see section 3 - parameter tuning). 

![Left cam](./examples/left_2018_04_30_19_51_00_029.jpg "Left camera")
![Center cam](./examples/center_2018_04_30_19_51_00_029.jpg "Left camera")
![Right cam](./examples/right_2018_04_30_19_51_00_029.jpg "Left camera")

Adding these camera images brought a multiplier of 3 to my image dataset.

To augment the data sat, I also flipped every image, to remove any bias that might occur from too many image sets in one or the other direction. Of course, the steering information was reverted, too. 

I thought about adding more training data by recovering the vehicle from the side of the road. Further, I planned to introduce training data from the second track. But since the model was so effective in driving the car autonomously already, I decided to skip this step.

After the collection process, I had 31656 number of data points. Data augmentation increased this number to 63312 images. 

I shuffled the data with a seed to obtain a reproducible validation test set. I picked a size of 20% for the validation test set and kept 80% for training the model. The validation set helped determine if the model was over or under fitting. The ideal number of epochs was Z as evidenced by running the network for 10 epochs. In almost all cases, the best solution was found after two to three epochs, probably due to the large dataset that was applied. 

The final model was working really well on the test track. Even at a speed of 30mph, it was driving smoothly without shaking too much. After one lap, I turned the vehicle around manually. Even for the other direction, autonomous mode was working perfectly!

![Model v10](./examples/record10_30mph.gif "Model v10: Final model")

Since the last GIF is not always loading properly (filesize too large... 9M), I created an additional video file showing how the model is running and how the system behaves. It is stored on [YouTube](https://www.youtube.com/watch?v=kMGlbVB5Nu4). Enjoy watching!

[![YouTube Video](https://img.youtube.com/vi/kMGlbVB5Nu4/0.jpg)](https://www.youtube.com/watch?v=kMGlbVB5Nu4)

It was again a really nice and interesting project to work on.

For me it was especially surprising, that a model was obtained only by providing images and a steering wheel information!


HISTORY.md
* record1: Dataset only from Udacity, 2 epochs


drive.py
import sys
print(sys.executable)
print(sys.version)

import argparse
import base64
from datetime import datetime
import os
import shutil
import numpy as np
import socketio
import eventlet
import eventlet.wsgi
from PIL import Image
from flask import Flask
from io import BytesIO

from keras.models import load_model
import h5py
from keras import __version__ as keras_version

sio = socketio.Server()
app = Flask(__name__)
model = None
prev_image_array = None


class SimplePIController:
    def __init__(self, Kp, Ki):
        self.Kp = Kp
        self.Ki = Ki
        self.set_point = 0.
        self.error = 0.
        self.integral = 0.

    def set_desired(self, desired):
        self.set_point = desired

    def update(self, measurement):
        # proportional error
        self.error = self.set_point - measurement

        # integral error
        self.integral += self.error

        return self.Kp * self.error + self.Ki * self.integral


controller = SimplePIController(0.1, 0.002)
set_speed = 30
controller.set_desired(set_speed)


@sio.on('telemetry')
def telemetry(sid, data):
    if data:
        # The current steering angle of the car
        steering_angle = data["steering_angle"]
        # The current throttle of the car
        throttle = data["throttle"]
        # The current speed of the car
        speed = data["speed"]
        # The current image from the center camera of the car
        imgString = data["image"]
        image = Image.open(BytesIO(base64.b64decode(imgString)))
        image_array = np.asarray(image)
        steering_angle = float(model.predict(image_array[None, :, :, :],batch_size=1))

        throttle = controller.update(float(speed))

        print(steering_angle, throttle)
        send_control(steering_angle, throttle)

        # save frame
        if args.image_folder != '':
            timestamp = datetime.utcnow().strftime('%Y_%m_%d_%H_%M_%S_%f')[:-3]
            image_filename = os.path.join(args.image_folder, timestamp)
            image.save('{}.jpg'.format(image_filename))
    else:
        # NOTE: DON'T EDIT THIS.
        sio.emit('manual', data={}, skip_sid=True)


@sio.on('connect')
def connect(sid, environ):
    print("connect ", sid)
    send_control(0, 0)


def send_control(steering_angle, throttle):
    sio.emit(
        "steer",
        data={
            'steering_angle': steering_angle.__str__(),
            'throttle': throttle.__str__()
        },
        skip_sid=True)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Remote Driving')
    parser.add_argument(
        'model',
        type=str,
        help='Path to model h5 file. Model should be on the same path.'
    )
    parser.add_argument(
        'image_folder',
        type=str,
        nargs='?',
        default='',
        help='Path to image folder. This is where the images from the run will be saved.'
    )
    args = parser.parse_args()

    # check that model Keras version is same as local Keras version
    f = h5py.File(args.model, mode='r')
    model_version = f.attrs.get('keras_version')
    keras_version = str(keras_version).encode('utf8')

    if model_version != keras_version:
        print('You are using Keras version ', keras_version,
              ', but the model was built using ', model_version)

    model = load_model(args.model)

    if args.image_folder != '':
        print("Creating image folder at {}".format(args.image_folder))
        if not os.path.exists(args.image_folder):
            os.makedirs(args.image_folder)
        else:
            shutil.rmtree(args.image_folder)
            os.makedirs(args.image_folder)
        print("RECORDING THIS RUN ...")
    else:
        print("NOT RECORDING THIS RUN ...")

    # wrap Flask application with engineio's middleware
    app = socketio.Middleware(sio, app)

    # deploy as an eventlet WSGI server
    eventlet.wsgi.server(eventlet.listen(('', 4567)), app)

    for num in [1,2,3,4]:
        print()


video.py
import moviepy.editor as mpy
import argparse
import os


IMAGE_EXT = ['jpeg', 'png', 'jpg']


def main():
    parser = argparse.ArgumentParser(description='Create driving video.')
    parser.add_argument(
        'image_folder',
        type=str,
        default='',
        help='Path to image folder. The video will be created from these images.'
    )
    parser.add_argument(
        '--fps',
        type=int,
        default=60,
        help='FPS (Frames per second) setting for the video.')
    args = parser.parse_args()

    #convert file folder into list firltered for image file types
    image_list = sorted([os.path.join(args.image_folder, image_file)
                        for image_file in os.listdir(args.image_folder)])

    image_list = [image_file for image_file in image_list if os.path.splitext(image_file)[1][1:].lower() in IMAGE_EXT]

    #two methods of naming output video to handle varying environemnts
    video_file_1 = args.image_folder + '/' + args.image_folder + '.mp4'
    gif_file_1 = args.image_folder + '/' + args.image_folder + '.gif'
    video_file_2 = args.image_folder + '/' + 'output_video.mp4'
    gif_file_2 = args.image_folder + '/' + 'output_video.gif'

    print("\nCreating video {}, FPS={}".format(args.image_folder, args.fps))
    print("Creating GIF {}, FPS=4 with every 15th image file".format(args.image_folder))

    clip = mpy.ImageSequenceClip(image_list, fps=args.fps)
    clip_gif = mpy.ImageSequenceClip(image_list[0::15], fps=4)

    try:
        clip.write_videofile(video_file_1)
        clip_gif.write_gif(gif_file_1)
    except:
        clip.write_videofile(video_file_2)
        clip_gif.write_gif(gif_file_2)

if __name__ == '__main__':
    main()


model.py
import csv
import cv2
import numpy as np
import ntpath
import time
import pickle
import sklearn
from tqdm import tqdm
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Flatten, Dense, Lambda, Cropping2D, Convolution2D, Dropout


######################################################################################
### Settings
# use_datasets = [0,1,2]  # Array to select which datasets to process
use_datasets = [3]      # Array to select which datasets to process
nb_epochs = 7           # Number of epochs for neural network training
batch_sz = 32           # Batch size for neural network
test_sz = 0.20          # Fraction of images to use for test set
steer_corr = 0.00       # Steering correction value (left, right camera)
dropout_keep_rate = 0.7 # Dropout rate to keep neurons

### List of available datasets
datasets = np.array([   'data/Udacity/driving_log.csv', \
                        'data/T1_Regular/driving_log.csv', \
                        'data/T1_OtherDir/driving_log.csv', \
                        'data/driving_log.csv'], \
                        dtype='str')
datasets = datasets[use_datasets]

################################################################################
### Function to read in image files and store them as pickle file
def read_csvs(datasets, steer_corr):
    ### Open CSV files of provided datasets
    images = []
    measurements = []
    for csvpath in datasets:
        with open(csvpath) as csvfile:
        # with open('data/Udacity/driving_log.csv') as csvfile:
            print('Processing file "',csvpath,'"... ',sep='',end='')
            reader = csv.reader(csvfile, skipinitialspace=True)
            j = 0
            for line in reader:
                if j == 0:
                    lines_header = line
                    j = j + 1
                    continue
                else:
                    # Update image file paths for center, left and right cam
                    line[0] = ntpath.split(csvpath)[0] + \
                                    '/IMG/'+ntpath.split(line[0])[-1] # center
                    line[1] = ntpath.split(csvpath)[0] + \
                                    '/IMG/'+ntpath.split(line[1])[-1] # left
                    line[2] = ntpath.split(csvpath)[0] + \
                                    '/IMG/'+ntpath.split(line[2])[-1] # right
                    # Add image path information
                    images.extend([line[0],line[1],line[2]])
                    # Add steering angle information
                    measurements.extend([
                        float(line[3]),
                        float(line[3])+steer_corr,
                        float(line[3])-steer_corr])
            print('DONE!')
            print('  Total amount of datasets is now at',len(images),
                'and',len(measurements),'steering infos')
    return images, measurements
################################################################################
### Read in files
images, steer = read_csvs(datasets, steer_corr)
### Split datasets between "train" and "test" dataset
images_train, images_test, steer_train, steer_test = train_test_split(
                                                            images,
                                                            steer,
                                                            test_size=test_sz,
                                                            random_state=42)
################################################################################
### Keras Neural Network
# Print out
print("\nStarting Keras")
print("  'X_train' and 'y_train' with {} elements\n".format(len(images_train)))
print("  'X_test'  and 'y_test'  with {} elements\n".format(len(images_test)))
################################################################################
### Define generator function
# Generator is used to read in image files during batch creation. Reason for
# introducing this feature was memory limitations: model is now able to process
# as many image files as needed!
def generator(datasets, steer, batch_size):
    num_samples = len(datasets)
    while 1: # Loop forever so the generator never terminates
        datasets, steer = sklearn.utils.shuffle(datasets, steer)
        for offset in range(0, num_samples, int(batch_size/2)):
            # Half batch size is used, because images are augmented with
            # vertically rotated pictures --> multiplier of 2, but half as
            # many pictures should be read in one batch
            batch_datasets = datasets[offset:offset+int(batch_size/2)]
            batch_steers   = steer[offset:offset+int(batch_size/2)]

            images = []
            for batch_dataset in batch_datasets:
                image = cv2.cvtColor(cv2.imread(batch_dataset),cv2.COLOR_BGR2RGB)
                image_aug = cv2.flip(image,1)
                images.append(image)
                images.append(image_aug)
            angles = []
            for batch_steer in batch_steers:
                angles.extend([float(batch_steer),float(batch_steer)*-1])

            X_batch = np.array(images,dtype='float64')
            y_batch = np.array(angles,dtype='float64')

            yield sklearn.utils.shuffle(X_batch, y_batch)
################################################################################
### Create generators for train and test datasets
train_generator = generator(images_train, steer_train, batch_size=batch_sz)
test_generator = generator(images_test, steer_test, batch_size=batch_sz)
### Define model layout (based on NVidia model as indicated in video
# (URL: https://devblogs.nvidia.com/deep-learning-self-driving-cars/)
# Original NVidia input size: 66,200,3
model = Sequential()
# Perform normalization of image data
model.add(Lambda(lambda x: x / 255.0 - 0.5, input_shape = (160,320,3)))
# Crop images: skip upper 70 and lower 25 pixels
model.add(Cropping2D(cropping=((70,25), (0,0))))  # remaining size: 65,320,3
# Several convolution layers, following NVidia
model.add(Convolution2D(24,5,5,subsample=(2,2),activation="relu"))
model.add(Convolution2D(36,5,5,subsample=(2,2),activation="relu"))
model.add(Convolution2D(48,5,5,subsample=(2,2),activation="relu"))
model.add(Convolution2D(64,3,3,activation="relu"))
model.add(Convolution2D(64,3,3,activation="relu"))
model.add(Flatten())
model.add(Dropout(dropout_keep_rate))
model.add(Dense(100))
model.add(Dense(50))
model.add(Dense(10))
model.add(Dense(1))
### Model optimization strategy
model.compile(loss='mse', optimizer='adam')
### Execute model
model.fit_generator(train_generator,
                    samples_per_epoch=len(images_train)*2, # x2: augmentation
                    nb_epoch=nb_epochs,
                    validation_data=test_generator,
                    nb_val_samples=len(images_test)*2) # x2: augmentation
### Save model
print("\n Saving neural network model as 'model.h5'...",end='')
model.save('model.h5')
print("DONE!")
print("Script finished.")
quit()
################################################################################
